### PR TITLE
[LTO] Allow target-specific module splittting

### PR DESCRIPTION
**Pull request by Pierre-vh**
Allow targets to implement custom module splitting logic for --lto-partitions, see #89245

https://discourse.llvm.org/t/rfc-lto-target-specific-module-splittting/77252

### CODE REVIEW

**Review by teresajohnson — COMMENTED** [review]
@pcc who implemented split LTO code gen

I am not a user of this functionality, but it seems reasonable to me. It would be good to see the companion PR with the use case as well when reviewing.

File: llvm/tools/llvm-split/llvm-split.cpp [hunk 1]
```diff
@@ -61,28 +100,41 @@ int main(int argc, char **argv) {
...
+    SplitModule(*M, NumOutputs, HandleModulePart, PreserveLocals);
+    return 0;
+  }
+
+  if (PreserveLocals) {
+    errs() << "warning: -preserve-locals has no effect when using "
+              "TargetMachine::splitModule\n";
+  }
+
+  if (!TM->splitModule(*M, NumOutputs, HandleModulePart)) {
```
[the remark this answers is not in the document]
Comment (Pierre-vh) [reply llvm/tools/llvm-split/llvm-split.cpp]: Done

**Comment by Pierre-vh** [comment]
#89245 is a target-specific implementation of module splitting, which uses this system.

File: llvm/tools/llvm-split/llvm-split.cpp [hunk 2]
```diff
@@ -47,12 +53,45 @@ static cl::opt<bool>
...
+static cl::opt<std::string>
+    MCPU("mcpu", cl::desc("Target CPU, ignored if -mtarget is not used"),
+         cl::value_desc("cpu"), cl::cat(SplitCategory));
+
 int main(int argc, char **argv) {
+  InitLLVM X(argc, argv);
+
+  // NOTE: If mtarget is not present we could avoid initializing targets to save
+  // time, but this is a testing tool and it's likely not worth the added
+  // complexity.
```
Comment (teresajohnson) [inline llvm/tools/llvm-split/llvm-split.cpp:72]: How much complexity is there? Isn't it just a simple if check like the one below at line 82? Probably more straightforward to just add that check if so.

**Review by teresajohnson — COMMENTED** [review]
If this is ready for a full review please remove "[RFC]" from the title. Also, can you add a basic llvm-split based test? Which I guess should just ensure that the fallback splitting works since there is no target implementation in this PR.

**Review by teresajohnson — APPROVED** [review]

**Comment by dtcxzyw** [comment]
@Pierre-vh This patch breaks our downstream build (with GCC 10.2.1):

```
FAILED: bin/llvm-split 
: && /usr/bin/c++ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -w -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -Wl,-rpath-link,/home/dtcxzyw/llvm-build/./lib  -Wl,--gc-sections tools/llvm-split/CMakeFiles/llvm-split.dir/llvm-split.cpp.o -o bin/llvm-split  -Wl,-rpath,"\$ORIGIN/../lib:/home/dtcxzyw/llvm-build/lib:"  lib/libLLVMX86AsmParser.so.19.0git  lib/libLLVMRISCVAsmParser.so.19.0git  lib/libLLVMX86CodeGen.so.19.0git  lib/libLLVMRISCVCodeGen.so.19.0git  lib/libLLVMX86Desc.so.19.0git  lib/libLLVMRISCVDesc.so.19.0git  lib/libLLVMX86Info.so.19.0git  lib/libLLVMRISCVInfo.so.19.0git  -lpthread  lib/libLLVMCodeGen.so.19.0git  lib/libLLVMTarget.so.19.0git  lib/libLLVMBitWriter.so.19.0git  lib/libLLVMTransformUtils.so.19.0git  lib/libLLVMIRReader.so.19.0git  lib/libLLVMMC.so.19.0git  lib/libLLVMCore.so.19.0git  lib/libLLVMSupport.so.19.0git  -Wl,-rpath-link,/home/dtcxzyw/llvm-build/lib && :
/usr/bin/ld: tools/llvm-split/CMakeFiles/llvm-split.dir/llvm-split.cpp.o: undefined reference to symbol '_ZN4llvm6TripleC1ERKNS_5TwineE'
/usr/bin/ld: /home/dtcxzyw/llvm-build/./lib/libLLVMTargetParser.so.19.0git: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
```

**Comment by Pierre-vh** [comment]
> @Pierre-vh This patch breaks our downstream build (with GCC 10.2.1):
> 
> ```
> FAILED: bin/llvm-split 
> : && /usr/bin/c++ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -w -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -Wl,-rpath-link,/home/dtcxzyw/llvm-build/./lib  -Wl,--gc-sections tools/llvm-split/CMakeFiles/llvm-split.dir/llvm-split.cpp.o -o bin/llvm-split  -Wl,-rpath,"\$ORIGIN/../lib:/home/dtcxzyw/llvm-build/lib:"  lib/libLLVMX86AsmParser.so.19.0git  lib/libLLVMRISCVAsmParser.so.19.0git  lib/libLLVMX86CodeGen.so.19.0git  lib/libLLVMRISCVCodeGen.so.19.0git  lib/libLLVMX86Desc.so.19.0git  lib/libLLVMRISCVDesc.so.19.0git  lib/libLLVMX86Info.so.19.0git  lib/libLLVMRISCVInfo.so.19.0git  -lpthread  lib/libLLVMCodeGen.so.19.0git  lib/libLLVMTarget.so.19.0git  lib/libLLVMBitWriter.so.19.0git  lib/libLLVMTransformUtils.so.19.0git  lib/libLLVMIRReader.so.19.0git  lib/libLLVMMC.so.19.0git  lib/libLLVMCore.so.19.0git  lib/libLLVMSupport.so.19.0git  -Wl,-rpath-link,/home/dtcxzyw/llvm-build/lib && :
> /usr/bin/ld: tools/llvm-split/CMakeFiles/llvm-split.dir/llvm-split.cpp.o: undefined reference to symbol '_ZN4llvm6TripleC1ERKNS_5TwineE'
> /usr/bin/ld: /home/dtcxzyw/llvm-build/./lib/libLLVMTargetParser.so.19.0git: error adding symbols: DSO missing from command line
> collect2: error: ld returned 1 exit status
> ```

adb25477fb41dc42fc1abef75039dd7b60c05206

**Comment by dtcxzyw** [comment]
> > @Pierre-vh This patch breaks our downstream build (with GCC 10.2.1):
> > ```
> > FAILED: bin/llvm-split 
> > : && /usr/bin/c++ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -w -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -Wl,-rpath-link,/home/dtcxzyw/llvm-build/./lib  -Wl,--gc-sections tools/llvm-split/CMakeFiles/llvm-split.dir/llvm-split.cpp.o -o bin/llvm-split  -Wl,-rpath,"\$ORIGIN/../lib:/home/dtcxzyw/llvm-build/lib:"  lib/libLLVMX86AsmParser.so.19.0git  lib/libLLVMRISCVAsmParser.so.19.0git  lib/libLLVMX86CodeGen.so.19.0git  lib/libLLVMRISCVCodeGen.so.19.0git  lib/libLLVMX86Desc.so.19.0git  lib/libLLVMRISCVDesc.so.19.0git  lib/libLLVMX86Info.so.19.0git  lib/libLLVMRISCVInfo.so.19.0git  -lpthread  lib/libLLVMCodeGen.so.19.0git  lib/libLLVMTarget.so.19.0git  lib/libLLVMBitWriter.so.19.0git  lib/libLLVMTransformUtils.so.19.0git  lib/libLLVMIRReader.so.19.0git  lib/libLLVMMC.so.19.0git  lib/libLLVMCore.so.19.0git  lib/libLLVMSupport.so.19.0git  -Wl,-rpath-link,/home/dtcxzyw/llvm-build/lib && :
> > /usr/bin/ld: tools/llvm-split/CMakeFiles/llvm-split.dir/llvm-split.cpp.o: undefined reference to symbol '_ZN4llvm6TripleC1ERKNS_5TwineE'
> > /usr/bin/ld: /home/dtcxzyw/llvm-build/./lib/libLLVMTargetParser.so.19.0git: error adding symbols: DSO missing from command line
> > collect2: error: ld returned 1 exit status
> > ```
> 
> [adb2547](https://github.com/llvm/llvm-project/commit/adb25477fb41dc42fc1abef75039dd7b60c05206)

Thank you!

**Pull request merged**